src/hotspot/share/classfile/classFileParser.cpp

```diff
@@ -5117,8 +5146,38 @@ void ClassFileParser::fill_instance_klass(InstanceKlass* ik,
   check_methods_for_intrinsics(ik, methods);
 
   // Fill in field values obtained by parse_classfile_attributes
-  if (_parsed_annotations->has_any_annotations()) {
+  if (_parsed_annotations->has_any_annotations())
     _parsed_annotations->apply_to(ik);
+
+  // Check the aot initialization safe status.
+  // @AOTSafeClassInitializer is used only to support ahead-of-time initialization of classes
+  // in the AOT assembly phase.
+  if (ik->has_aot_safe_initializer() && CDSConfig::is_initing_classes_at_dump_time()) {
+    // If a type is included in the tables inside can_archive_initialized_mirror(), we require that
+    //   - all super classes must be included
+    //   - all super interfaces that have <clinit> must be included.
+    // This ensures that in the production run, we don't run the <clinit> of a supertype but skips
+    // ik's <clinit>.
+    if (_super_klass != nullptr) {
+      if (!_super_klass->has_aot_safe_initializer()) {
+        classfile_parse_error("aot-initialization safe %s requires an aot-initialization safe super class", CHECK);
+      }
+    }
+
+    int len = _transitive_interfaces->length();
+    for (int i = 0; i < len; i++) {
+      InstanceKlass* intf = _transitive_interfaces->at(i);
+      if (intf->class_initializer() != nullptr) {
+        if (!intf->has_aot_safe_initializer()) {
```

Comment:
I think this is better, so you don't need to annotate a supertypes that have no `<clinit>`

```
if (intf->class_initializer() != nullptr && !intf->has_aot_safe_initializer()) {
```

Reply:
Also quick question: Should I use `_transitive_interfaces` or `_local_interfaces`?

Reply:
local_interfaces is fine, because the interfaces implemented by the super classes would have been checked when the super classes were loaded.

BTW, the error message should include the name of both this class and the supertype:

```
classfile_parse_error("AOTSafeClassInitlaizer annotation is required for supertype %s of %s", ...
```

And the checks can be refactored in a separate function to avoid repetition.

```
if (_super_klass != nullptr) {
  check_aot_safe_initializer(ik, _super_klass, CHECK);
}
int len = _local_interfaces->length();
for (int i = 0; i < len; i++) {
  check_aot_safe_initializer(ik, _local_interfaces->at(i), CHECK);
}
```

Reply:
I don't want to introduce a new declaration in the header for this function - to shrink the size, I think I can use `guarantee_property` instead.